### ISSUE TITLE
openstack_compute_instance_v2 not refreshing metadata and updating metadata wihtout displaying changes

### ISSUE THREAD

**Issue by NilsBusche** [terraform-providers/terraform-provider-openstack#875]
### Terraform Version
Terraform v0.12.9

### Affected Resource(s)
openstack_compute_instance_v2

### Terraform Configuration Files
```hcl
resource "openstack_compute_instance_v2" "instance" {
  name      = "tag_test"
  image_id  = "842ee4ac-6caa-449b-866a-4ffa69e90be2"
  flavor_id = "38b76bdd-746d-40f9-878b-450304001cd3"

  network {
    port = openstack_networking_port_v2.nic_0.id
  }

  metadata = {
     tag1 = "test",
     tag2 = "test"
  }
}

```

### Expected Behavior
If you update the metadata by using openstack methods, the next `terraform apply` should display changes:
`openstack server set --property tag2=new tag_test`
`terraform apply`

Expected output:
```hcl
~ metadata            = {
            "tag1" = "test"
          ~ "tag2" = "new" -> "test"
        }
```

### Actual Behavior
If you update metadata with openstack methods, terraform does not recognize this and is not displaying changes. If you update then another key-value-pair now using terraform, it will overwrite the before updated key-value-pair without displaying this change to you.

### Steps to Reproduce
Update metadata:
`openstack server set --property tag2=new tag_test`

Check (is updated):
```
> openstack server show tag_test -c properties -f value
tag1='test', tag2='new'
```

Run `terraform apply`:
```hcl
openstack_compute_instance_v2.instance: Refreshing state... [id=5abf4ce4-c5bd-4b6e-8df3-8e629a411d4f]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

Check (nothing happened):
```
> openstack server show tag_test -c properties -f value
tag1='test', tag2='new'
```

Update metadata in terraform configuration:
```hcl
metadata = {
     tag1 = "new",
     tag2 = "test"
  }
```

Run `terraform apply`:
```hcl
[...]
      ~ metadata            = {
          ~ "tag1" = "test" -> "new"
            "tag2" = "test"
        }
[...]
```

Check (tag1 was updated but tag2 also although not displayed in changes):
```
> openstack server show tag_test -c properties -f value
tag1='new', tag2='test'
```

**Comment by jtopjian** [comment]
@NilsBusche For metadata set outside of Terraform, use the `all_metadata` attribute.

**Comment by NilsBusche** [comment]
> @NilsBusche For metadata set outside of Terraform, use the `all_metadata` attribute.

Okay, this makes sense to work with key-value-pairs that are not in the terraform configuration.

But the bug I've described above is the following:

Assume you set some key-value-pairs in the metadata of a instance using terraform. Then someone changes one of these values outside of terraform because he does not know that there is a terraform integration. Now terraform should recognize this in a `terraform plan` and a `terraform apply`, so that it will change it back to the value how it was set in the terraform configuration files.

This is one of the benefits of terraform, that it is comparing between how the configuration is at the moment and how it should be like described in the terraform configuration files.

**Comment by jtopjian** [comment]
@NilsBusche Agreed. Unfortunately that is not supported at this time for instance metadata specifically. Other areas support this, but metadata (and metadata-like behaviors in a few other areas of this provider) is a bit of an edge case.

Consider the following scenarios:

1. An instance created in Terraform with metadata supplied by the user.
2. An instance created in Terraform with metadata provided by the OpenStack environment.
3. Updating an instance's metadata through Terraform.
4. Updating an instance's metadata outside of Terraform.

Scenario 1 and 3 are the easiest to imagine, so we'll skip them.

Scenarios 2 and 4 are more difficult. If the OpenStack environment sets metadata on the instance, how should Terraform reconcile that? This requires reconciling two sources of truth to determine a third converged truth. Is it possible to apply one default policy to account for everyone's use-case? Additionally, what if someone _removes_ metadata outside of Terraform. How does reconciliation happen now?

It's because of this difficulty that not much work has been done to resolve this. The team here is made up of volunteers, so it's a matter of balancing personal time. If you're able to work on this issue, I'm sure they'd be appreciative!

**Comment by NilsBusche** [comment]
Thanks for this detailed explanation. Now I understand the problem. I only was wondering about this behaviour because other terraform providers like AWS or Azure simply do not allow doing anything outside of terraform. Those changes would be simply overwritten by terraform. 
Do you know if it was a explicit design decision for the openstack provider allowing metadata changes outside of terraform?

**Comment by buec011** [comment]
IMHO: I work with AWS, Azure, OCI, vCD and in near future with the google terraform provider.
All of them work like they are the only source of truth and I assumed it for openstack as well.

If you want to keep metadata provided by Openstack (case 2) or other third party (case 4) you need to add them to your template / module / tfvars ...
Example: We deploy in AWS instances with terraform in an account not managed by us / terraform. In that account graffiti monkey is active (https://github.com/Answers4AWS/graffiti-monkey). It adds tags from the instance to volumes. So we need to add the tags in our template, that was not a problem. With terraform 0.12 you can even ignore tags (e.g. if the tag will be changed every day for whatever reason or the customer wants to write the next reboot time in it, etc) (see https://www.terraform.io/docs/configuration/resources.html , chapter lifecycle: Lifecycle Customizations )

So I suggest the following for scenarios 2 and 4: It will be overwritten by terraform, the one and only truth if you use terraform .

**Comment by buec011** [comment]
PS: It looks like the developer followed the same idea as in 
https://github.com/terraform-providers/terraform-provider-openstack/issues/552
And maybe the solution could be similar: Allow `all_metadata` as input, that enforce the metadata, or a `enforce_metadata=true` option?

**Comment by jtopjian** [comment]
> If you want to keep metadata provided by Openstack (case 2) or other third party (case 4) you need to add them to your template / module / tfvars ...

If I understand correctly, you mean that once you see that metadata has been added out-of-band of Terraform, you modify your Terraform configurations to account for them? That's certainly one solution.

> So I suggest the following for scenarios 2 and 4: It will be overwritten by terraform, the one and only truth if you use terraform

and

> It looks like the developer followed the same idea as in

Correct.

I did much of the work on this OpenStack provider over the past few years, though I no longer work on it now. I asked for advice from other Terraform developers for the best solutions _at the time_. 

There are pros and cons to every solution, and it's quite possible that consensus during discussions leaned toward the current solution and consensus another day would have leaned toward another solution. 🤷‍♂ 

> enforce_metadata=true

This is something that could be added to the current `openstack_compute_instance_v2` resource and not affect prior behavior.

Also note that `openstack_compute_instance_v2` is a big and complicated resource. I once had grand plans for a [new resource](https://github.com/terraform-providers/terraform-provider-openstack/issues/457) but unfortunately don't have the time to look into it nowadays.

**Comment by buec011** [comment]
> If I understand correctly, you mean that once you see that metadata has been added out-of-band of Terraform, you modify your Terraform configurations to account for them? 

Exactly. 

I read many bugs for different provider yet. My impression is that it is typical that one have a CI/CD pipline similar to
- Save a terraform plan
- Review
- Approve
- terraform deploy

Ofcourse it will be tricky here: On initial deployment of an instance terraform may not know the tags that has been added by an openstack based cloud. You will get aware of it during next terraform plan, which wants to remove those. But if you have a review step, you may get aware of it. And after first instance, you know your openstack cloud and can adjust templates.

In addition, we have a "nightly check" / monitoring: We run `terraform plan` every night and raise a ticket if infrastructure does not match terraform. But this use case needs to enforce everything and should not allow tasks outside of terraform. For other provider that I know, this is the case.
